docs/source/ContributionGuide.rst

```diff
@@ -65,10 +65,119 @@ be submitted as a github Pull Request. This process tries to ensure that every
+
+- Golden tip: never work on local branch master. Instead, create a new
+  local branch and checkout to it::
+
+    $ git checkout -b my_new_local_branch
+
+- Code and then commit your changes::
+
+    $ git add new-file.py
+    $ git commit -a -s
+
+- Write a good commit message, pointing motivation, issues that you're
+  addressing. Usually you should try to explain 3 points in the commit
+  message: motivation, approach and effects::
+
+    header          <- Limited to 50 characters. No period.
+                    <- Blank line
+    message         <- Any number of lines, limited to 72 characters per line.
+                    <- Blank line
+    Reference:      <- External references, one per line (issue, trello, ...)
+    Signed-off-by:  <- Signature (created by git commit -s)
+
+- Make sure your code is working (install your version of avocado, test
+  your change, run ``make check`` to make sure you didn't introduce any
+  regressions).
+
+- Paste the ``job.log`` file content from the previous step in a pastebin
+  service, like fpaste.org.
+
+- Push your commit(s) to your fork::
+
+    $ git push origin my_new_local_branch
+
+- Create the Pull Request on github. Add the relevant information to the
+  Pull Request description.
+
+- In the Pull Request discussion page, comment with the link to the
+  ``job.log`` pastebin.
+
+- Check if your Pull Request passes the CI (travis). Your Pull Request
+  will probably be ignored until it's all green.
+
+Now you're waiting for feedback on github Pull Request page. Once you
+get some, join the discussion, answer the questions, make clear if you're
+going to change the code based on some review and, if not, why. New versions
+of your code should not be force-updated (unless explicitly requested by
+the code reviewer). Instead, you should:
+
+- Create a new branch out of your previous branch::
+
+    $ git checkout my_new_local_branch
+    $ git checkout -b my_new_local_branch_v2
+
+- Code, and amend the commit(s) and/or create new commits. If you have
+  more than one commit in the PR, you will probably need to rebase
+  interactively to amend the right commits. ``git cola`` or ``git citool``
+  can be handy here.
+
+- Rebase your local branch on top of upstream master::
+
+    $ git checkout master
+    $ git pull upstream master
+    $ git push
+    $ git checkout my_new_local_branch_v2
+    $ git rebase origin/master
+    (resolve merge conflicts, if any)
+
+- Push your changes::
+
+    $ git push origin my_new_local_branch_v2
+
+- Create a new Pull Request for this new branch. In the Pull Request
+  description, point the previous Pull Request and the changes the current
+  Pull Request introduced when compared to the previous Pull Request(s).
+
+- Close the previous Pull Request on github.
+
+After your PR gets merged, you can sync your local repository and your
+fork on github::
+
+    $ git checkout master
+    $ git pull upstream master
+    $ git push
```

Comment:
Sync from upstream to local master and then from local master to origin(fork) master. After that, the 3 master branches are in sync: upstream(project repo), origin(user fork) and local.